packages/@tailwindcss-upgrade/src/template/codemods/modernize-arbitrary-values.ts

```diff
@@ -0,0 +1,147 @@
+import SelectorParser from 'postcss-selector-parser'
+import type { Config } from 'tailwindcss'
+import { parseCandidate, type Candidate, type Variant } from '../../../../tailwindcss/src/candidate'
+import type { DesignSystem } from '../../../../tailwindcss/src/design-system'
+import { printCandidate } from '../candidates'
+
+export function modernizeArbitraryValues(
+  designSystem: DesignSystem,
+  _userConfig: Config,
+  rawCandidate: string,
+): string {
+  for (let candidate of parseCandidate(rawCandidate, designSystem)) {
+    let clone = structuredClone(candidate)
+    let changed = false
+
+    for (let [variant, parent] of variants(clone)) {
+      // Expecting an arbitrary variant
+      if (variant.kind !== 'arbitrary') continue
+
+      // Expecting a non-relative arbitrary variant
+      if (variant.relative) continue
+
+      let ast = SelectorParser().astSync(variant.selector)
+
+      // Expecting a single selector node
+      if (ast.nodes.length !== 1) continue
+
+      // Track whether we need to add a `*:` variant
+      let addChildVariant = false
+
+      // Handling a child combinator. E.g.: `[&>[data-visible]]` => `*:data-visible`
+      if (
+        // Only top-level, so `has-[&>[data-visible]]` is not supported
+        parent === null &&
+        // [&_>_[data-visible]]:flex
+        //  ^ ^ ^^^^^^^^^^^^^^
+        ast.nodes[0].length === 3 &&
+        ast.nodes[0].nodes[0].type === 'nesting' &&
+        ast.nodes[0].nodes[0].value === '&' &&
+        ast.nodes[0].nodes[1].type === 'combinator' &&
+        ast.nodes[0].nodes[1].value === '>' &&
+        ast.nodes[0].nodes[2].type === 'attribute'
+      ) {
+        ast.nodes[0].nodes = [ast.nodes[0].nodes[2]]
+        addChildVariant = true
```

Comment:
Only thing I can think of is that `*:data-visible:{utility}` is _technically_ something like `:where(& > [data-visible])` but it might be fine to ignore that detail.

If we're cool with that this all looks good.

Reply:
Thinking about this, I think the new value is _more_ correct because with just `[&>[data-visible]]` you run into more issues: https://play.tailwindcss.com/NPsMLWYpGZ

Reply:
I'm fine with making the change. Maybe should output a small note when we find it in case the old behavior is being relied on?

Reply:
I'd also be fine with doing that in a follow up PR honestly.

Reply:
That's an option. We technically already mention to verify the changes. What are your thoughts on this /cc @philipp-spiess @adamwathan

Reply:
Yeah that's fair. Maybe that's fine.

Reply:
Tricky one since the specificity is different, not sure. We probably can change it but trying to think of cases where it would matter. Should test in Catalyst which does a lot of this sort of thing.

Reply:
There are indeed cases in Catalyst where it would break if we do  this conversion sadly. We could go through Catalyst and make sure that this isn't an issue, but that doesn't solve it for our users.

I'll revert this particular change.

Reply:
Revert this back because we made `*` use `:is(…)` instead of `:where(…)`.